Trigger the loadTablets function when no primary is assigned to the shard

yoheimuta opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->

This PR ensures that the vtgate will route traffic to the new primary tablet without waiting for the `tabletRefreshInterval`.

This is to prevent situations where a new primary tablet for a shard has just been created and the `tableRefreshInterval` has not yet passed, causing an interruption where no primary is assigned to the shard.

## Related Issue(s)

<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

- Fixes #12624

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

yoheimuta commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+	time.Sleep(1 * time.Second)
+
+	// verify that the vtgate will recognize the new primary tablet
+	qr, _ := vtgateConn.ExecuteFetch("show vitess_tablets", 100, true)
```
If you run the test code with the slight modifications against the current main, the following output will be generated:

```
...
I0414 16:43:31.320373   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 TabletExternallyReparented healthcheck_test_cell-0000009999 (attempt 1 of 10)
I0414 16:43:31.472102   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 SetReadWrite healthcheck_test_cell-0000009999 (attempt 1 of 10)
    reparent_test.go:161: qr=[[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("REPLICA") VARCHAR("NOT_SERVING") VARCHAR("healthcheck_test_cell-0000001336") VARCHAR("localhost") VARCHAR("")] [VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("80-") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000001337") VARCHAR("localhost") VARCHAR("2023-04-14T07:41:48Z")]]
    reparent_test.go:162:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:162
                Error:          Not equal:
                                expected: 3
                                actual  : 2
                Test:           TestHealthCheckExternallyReparentNewTablet
                Messages:       wrong number of tablet records in healthcheck, expected 3 but had 2
    reparent_test.go:164:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:164
                Error:          Not equal:
                                expected: "PRIMARY"
                                actual  : "REPLICA"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -PRIMARY
                                +REPLICA
                Test:           TestHealthCheckExternallyReparentNewTablet
    reparent_test.go:165:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:165
                Error:          Not equal:
                                expected: "SERVING"
                                actual  : "NOT_SERVING"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -SERVING
                                +NOT_SERVING
                Test:           TestHealthCheckExternallyReparentNewTablet
    reparent_test.go:166:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:166
                Error:          Not equal:
                                expected: "healthcheck_test_cell-0000009999"
                                actual  : "healthcheck_test_cell-0000001336"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -healthcheck_test_cell-0000009999
                                +healthcheck_test_cell-0000001336
                Test:           TestHealthCheckExternallyReparentNewTablet
I0414 16:43:42.524622   87256 mysqlctl_process.go:156] Shutting down MySQL: 1336

I0414 16:43:54.685033   87256 mysqlctl_process.go:175] MySQL shutdown complete: 1336
I0414 16:43:54.764293   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 DeleteTablet healthcheck_test_cell-0000001336 (attempt 1 of 10)
    reparent_test.go:243: Deleted tablet: healthcheck_test_cell-0000001336
    reparent_test.go:175:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:175
                Error:          Received unexpected error:
                                target: healthcheck_test_ks.-80.primary: current keyspace is being resharded (errno 1105) (sqlstate HY000) during query: insert into customer(customer_id, email) values(2, 'dummy1')
                Test:           TestHealthCheckExternallyReparentNewTablet
--- FAIL: TestHealthCheckExternallyReparentNewTablet (93.68s)
```

yoheimuta replied:
When you run the same test code against the current branch:

```
...
reparent_test.go:225: Added tablet: healthcheck_test_cell-0000009999
I0414 17:05:59.561821   92542 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 TabletExternallyReparented healthcheck_test_cell-0000009999 (attempt 1 of 10)
I0414 17:05:59.755164   92542 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 SetReadWrite healthcheck_test_cell-0000009999 (attempt 1 of 10)
    reparent_test.go:161: qr=[ # NOTE: Inserted new lines manually
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000009999") VARCHAR("localhost") VARCHAR("2023-04-14T08:05:59Z")] 
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("REPLICA") VARCHAR("NOT_SERVING") VARCHAR("healthcheck_test_cell-0000006955") VARCHAR("localhost") VARCHAR("")] 
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("80-") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000006956") VARCHAR("localhost") VARCHAR("2023-04-14T08:04:46Z")]
]
...
```

```
...
I0414 17:05:59.811311   93642 tablet_health_check.go:110] HealthCheckUpdate(Serving State): tablet: healthcheck_test_cell-6955 (localhost) serving true => false for healthcheck_test_ks/-80 (PRIMARY) reason: healthCheck update
I0414 17:05:59.814976   93642 healthcheck.go:509] We will have no health data for the next new primary tablet after demoting the tablet: healthcheck_test_cell-6955 (localhost), so start loading tablets now
I0414 17:05:59.835891   93642 healthcheck.go:382] Adding tablet to healthcheck: alias:{cell:"healthcheck_test_cell" uid:9999} hostname:"localhost" port_map:{key:"grpc" value:6722} port_map:{key:"vt" value:6721} keyspace:"healthcheck_test_ks" shard:"-80" key_range:{end:"\x80"} type:PRIMARY mysql_hostname:"localhost" mysql_port:6723 primary_term_start_time:{seconds:1681459559 nanoseconds:687021000} db_server_version:"5.7.39" default_conn_collation:255
I0414 17:05:59.837507   93642 tablet_health_check.go:110] HealthCheckUpdate(Serving State): tablet: healthcheck_test_cell-9999 (localhost) serving false => true for healthcheck_test_ks/-80 (PRIMARY) reason: healthCheck update
...
```

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+
+	// verify output of `show vitess_tablets` and `INSERT` statement
+	vtgateConn, err := mysql.Connect(ctx, &vtParams)
+	require.Nil(t, err)
```
Nit, but `require.NoError(t, err)` is more widely used and more explicit.

yoheimuta replied:
Fixed by [f2c1c7d](https://github.com/vitessio/vitess/pull/12893/commits/f2c1c7d716ad1de724ad4d676c26213a8cf1a3ed).

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+
+// TestHealthCheckExternallyReparentNewTablet ensures that calling TabletExternallyReparented on a new tablet will switch the primary tablet
+// without having to wait for the tabletRefreshInterval.
+func TestHealthCheckExternallyReparentNewTablet(t *testing.T) {
```
Have you executed all of the new tests locally, at least 20 times in a row, to be sure it's not flaky?

yoheimuta replied:
I never experienced a test failure during development, but I hadn't executed it 20 times in a row until now. So I ran two trials (`./test_script.sh`), which failed on the ninth attempt each time. I assume it's due to the machine load.

Fixed by [db90a74](https://github.com/vitessio/vitess/pull/12893/commits/db90a743fed4fc3fd73646eb7136bd678a003f97). I confirmed that the code passed locally 20 times in a row. 
You can find the log: https://gist.github.com/yoheimuta/840f49dabdf4373fe8a91144e06590be.

<details>
<summary>test_script.sh</summary>

```bash
#!/bin/bash

# Set the log file name
logfile="test_results.log"

# Clean up the log file
rm -f "$logfile"

# Execute the test and clean-up commands 20 times
for i in {1..20}; do
  echo "Running test iteration $i" | tee -a "$logfile"

  # Test
  go test -v -count=1 -run "^TestHealthCheckExternallyReparentNewTablet$" vitess.io/vitess/go/test/endtoend/vtgate/tablet_healthcheck --alsologtostderr >> "$logfile" 2>&1
  exit_code=$?

  # Check the exit code and exit the loop if the test failed
  if [ $exit_code -ne 0 ]; then
    echo "Test failed on iteration $i, stopping the script" | tee -a "$logfile"
    break
  fi

  # Clean-up
  rm -f /var/folders/yj/0nkt3j750nz3x582dpf84r3w0000gn/T/endtoend.port
  rm -rf /Users/yoheimuta/.go/vitess/vtdataroot/vtroot_6701
  killall mysqld vtgate vttablet etcd2

  echo "Finished test iteration $i" | tee -a "$logfile"
  echo "-------------------------------------------------" | tee -a "$logfile"
done

echo "All test iterations completed" | tee -a "$logfile"
```
</details>

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+	return fmt.Sprintf("%s-%010d", cell, tabletUID)
+}
+
+func addTablet(t *testing.T, tabletUID int, tabletType string) *cluster.Vttablet {
```
Is there anything unique about this function? I feel like we have other utilities we could use here instead.

yoheimuta replied:
Can you tell me where I can find the common addTablet utility function? And, if it's not available, do you think it's a good idea to create a new utility function since I've referenced [the adjacent test file](https://github.com/vitessio/vitess/blob/8c3e37c28bec8f76e818402127e8d37addf77867/go/test/endtoend/vtgate/tablet_healthcheck_cache/correctness_test.go#L176) in the first place?

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+
+	// wait for mysqld to be ready
+	err = proc.Wait()
+	require.Nil(t, err)
```
Same note here and below about using NoError.

yoheimuta replied:
Fixed by [f2c1c7d](https://github.com/vitessio/vitess/pull/12893/commits/f2c1c7d716ad1de724ad4d676c26213a8cf1a3ed).

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+	require.Nil(t, err)
+
+	serving := tablet.VttabletProcess.WaitForStatus("SERVING", time.Duration(60*time.Second))
+	assert.Equal(t, serving, true, "Tablet did not become ready within a reasonable time")
```
Any reason we're using assert here rather than require?

yoheimuta replied:
Fixed by [3b5799d](https://github.com/vitessio/vitess/pull/12893/commits/3b5799d390923c581653d87a0ae1b619707b281f).
I'm not particular about it. The reason is that I just referred to [the existing test](https://github.com/vitessio/vitess/blob/8c3e37c28bec8f76e818402127e8d37addf77867/go/test/endtoend/vtgate/tablet_healthcheck_cache/correctness_test.go#L217).

mattlord commented on go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:
```diff
@@ -0,0 +1,243 @@
+	return tablet
+}
+
+func deleteTablet(t *testing.T, tablet *cluster.Vttablet) {
```
This also feels like something used throughout the tests and maybe we can avoid adding another copy.

yoheimuta replied:
I have the same question here: https://github.com/vitessio/vitess/pull/12893#discussion_r1179908022.

mattlord commented on go/vt/discovery/healthcheck.go:
```diff
@@ -228,6 +229,9 @@ type HealthCheck interface {
 	Unsubscribe(c chan *TabletHealth)
+
+	// LoadTabletsTrigger returns a channel that is used to inform when to load tablets.
+	LoadTabletsTrigger() chan struct{}
```
IMO, `GetLoadTabletsTrigger` is a better name.

yoheimuta replied:
It makes perfectly sense.
Changed by [a710157](https://github.com/vitessio/vitess/pull/12893/commits/a7101571a76869bb0fb0f248553ed1b4a9235630).

mattlord reviewed (approved):
It makes sense to me. We should have @GuptaManan100 review though.

I had some minor comments that we can discuss, but will approve now since we need to wait for Manan anyway.

yoheimuta commented:
@mattlord Thanks for the review! I've just finished responding to your comments.

GuptaManan100 reviewed (commented):
The changes look good to me! It is probably a good idea to get another member from @vitessio/query-serving to take a look.

I had one more question. Does this problem arise only with `TabletExternallyReparented` or also with `PlannedReparentShard`?

yoheimuta commented:
@GuptaManan100 Thanks for taking the time to review! 

> Does this problem arise only with TabletExternallyReparented or also with PlannedReparentShard?

That seems to be true, although I haven't seen it in production because our MySQL is managed externally.

When I ran the test case in this PR with slight modifications against the current main, I was able to reproduce the same problem with PlannedReparentShard. Here's the output that was generated:

```diff
diff --git a/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go b/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go
index dbc46bdda7..a9b8a173b3 100644
--- a/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go
+++ b/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go
@@ -141,16 +142,17 @@ func TestHealthCheckExternallyReparentNewTablet(t *testing.T) {

        // add a new tablet
        reparentTabletUID := 9999
-       reparentTabletType := "rdonly"
+       reparentTabletType := "replica"
        tablet := addTablet(t, reparentTabletUID, reparentTabletType)

        // promote the new tablet to the primary
-       err = clusterInstance.VtctlclientProcess.ExecuteCommand("TabletExternallyReparented", tablet.Alias)
+       err = clusterInstance.VtctlclientProcess.ExecuteCommand("PlannedReparentShard", "--", "--keyspace_shard", keyspaceName+"/"+shards[0], "--new_primary", tablet.Alias)
        require.NoError(t, err)

-        // update the new primary tablet to be read-write
-       err = clusterInstance.VtctlclientProcess.ExecuteCommand("SetReadWrite", tablet.Alias)
-       require.NoError(t, err)
```

```sh
# My Comment: FAILED
reparent_test.go:226: Added tablet: healthcheck_test_cell-0000009999
I0502 18:35:50.019849    2609 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 PlannedReparentShard -- --keyspace_shard healthcheck_test_ks/-80 --new_primary healthcheck_test_cell-0000009999 (attempt 1 of 10)
    reparent_test.go:163:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:163
                Error:          Not equal:
                                expected: 3
                                actual  : 2
                Test:           TestHealthCheckExternallyReparentNewTablet
                Messages:       wrong number of tablet records in healthcheck, expected 3 but had 2. Got result=&{[name:"Cell" type:VARCHAR charset:33 flags:1 name:"Keyspace" type:VARCHAR charset:33 flags:1 name:"Shard" type:VARCHAR charset:33 flags:1 name:"TabletType" type:VARCHAR charset:33 flags:1 name:"State" type:VARCHAR charset:33 flags:1 name:"Alias" type:VARCHAR charset:33 flags:1 name:"Hostname" type:VARCHAR charset:33 flags:1 name:"PrimaryTermStartTime" type:VARCHAR charset:33 flags:1] 0 0 
# My Comment: Inserted new lines manually
# My Comment: The vtgate has recognized the previous primary demotion but not the latest primary promotion, resulting in the -80 shard being without a primary tablet.
[
   [VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("REPLICA") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000007558") VARCHAR("localhost") VARCHAR("")]
   [VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("80-") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000007559") VARCHAR("localhost") VARCHAR("2023-05-02T09:34:39Z")]
]  2 }
    reparent_test.go:165:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:165
                Error:          Not equal:
                                expected: "PRIMARY"
                                actual  : "REPLICA"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -PRIMARY
                                +REPLICA
                Test:           TestHealthCheckExternallyReparentNewTablet
    reparent_test.go:167:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:167
                Error:          Not equal:
                                expected: "healthcheck_test_cell-0000009999"
                                actual  : "healthcheck_test_cell-0000007558"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -healthcheck_test_cell-0000009999
                                +healthcheck_test_cell-0000007558
                Test:           TestHealthCheckExternallyReparentNewTablet
I0502 18:36:04.772983    2609 mysqlctl_process.go:157] Shutting down MySQL: 7558
I0502 18:36:16.037105    2609 mysqlctl_process.go:176] MySQL shutdown complete: 7558
I0502 18:36:16.108869    2609 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 DeleteTablet healthcheck_test_cell-0000007558 (attempt 1 of 10)
    reparent_test.go:244: Deleted tablet: healthcheck_test_cell-0000007558

# My Comment: vtgate failed to route the `INSERT` statement to the new primary tablet
    reparent_test.go:176:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:176
                Error:          Received unexpected error:
                                target: healthcheck_test_ks.-80.primary: current keyspace is being resharded (errno 1105) (sqlstate HY000) during query: insert into customer(customer_id, email) values(2, 'dummy1')
                Test:           TestHealthCheckExternallyReparentNewTablet
--- FAIL: TestHealthCheckExternallyReparentNewTablet (95.31s)
FAIL
E0502 18:36:16.963298    2609 cluster_process.go:1032] Error in mysqlctl teardown wait: exit status 1
FAIL    vitess.io/vitess/go/test/endtoend/vtgate/tablet_healthcheck     125.399s
FAIL
```

When I ran this modified test code against the current branch:

```sh
# My Comment: PASSED
...
reparent_test.go:225: Added tablet: healthcheck_test_cell-0000009999
I0502 18:19:11.714629   95212 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 PlannedReparentShard -- --keyspace_shard healthcheck_test_ks/-80 --new_primary healthcheck_test_cell-0000009999 (attempt 1 of 10)
I0502 18:19:29.157709   95212 mysqlctl_process.go:157] Shutting down MySQL: 8431
I0502 18:19:41.530447   95212 mysqlctl_process.go:176] MySQL shutdown complete: 8431
I0502 18:19:41.635834   95212 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 DeleteTablet healthcheck_test_cell-0000008431 (attempt 1 of 10)
    reparent_test.go:243: Deleted tablet: healthcheck_test_cell-0000008431
--- PASS: TestHealthCheckExternallyReparentNewTablet (99.66s)
PASS
E0502 18:19:43.225517   95212 cluster_process.go:1032] Error in mysqlctl teardown wait: exit status 1
ok      vitess.io/vitess/go/test/endtoend/vtgate/tablet_healthcheck     130.870s
```

harshit-gangal commented:
The changes look good. 
But, have you tried enabling vtgate buffering in production and see if that helps your use case?

yoheimuta commented:
@harshit-gangal Thank you for the review!

> But, have you tried enabling vtgate buffering in production 

Nice point. Yes, we let the vitess-operator set the default parameters for [vtgate buffering](https://vitess.io/docs/14.0/reference/features/vtgate-buffering/#vtgate-flags-to-enable-buffering) as follows:

```sh
$ kubectl describe pods/cocalero-api-asianortheast1a-vtgate-6085c9ff-676897cd9c-84ppl | grep buffer
      --buffer_max_failover_duration=10s
      --buffer_min_time_between_failovers=20s
      --buffer_size=1000
      --enable_buffer=true
```

> and see if that helps your use case?

Partially. The feature can be helpful if the disruption is brief and ceases quickly. However, in our previous incident, the disruption lasted for 15 seconds, so the feature ended up draining all the buffered requests after 10 seconds had elapsed. Therefore, the feature doesn't help in such cases.

You can find the logs generated during the reproducible step, where the `tablet_refresh_interval` remains the default 1 minute, here: https://gist.github.com/yoheimuta/c678cb7e5ea3453abe6ede3475e31e56#file-vtgate-log-txt-L31.
